examples/step-44/step-44.cc

```diff
@@ -2163,6 +2106,48 @@ namespace Step44
               fe.system_to_component_index(i).first;
             const unsigned int i_group = fe.system_to_base_index(i).first.first;
 
+            // We first compute the contributions
+            // from the internal forces.  Note, by
+            // definition of the rhs as the negative
+            // of the residual, these contributions
+            // are subtracted.
+            if (i_group == u_dof)
+              data.cell_rhs(i) -= (symm_grad_Nx[i] * tau) * JxW;
+            else if (i_group == p_dof)
+              data.cell_rhs(i) -= N[i] * (det_F - J_tilde) * JxW;
+            else if (i_group == J_dof)
+              data.cell_rhs(i) -= N[i] * (dPsi_vol_dJ - p_tilde) * JxW;
+            else
+              Assert(i_group <= J_dof, ExcInternalError());
+
+            // Before we go into the inner loop, we have one final chance to
+            // introduce some optimizations. We've already taken into account
+            // the symmetry of the system, and we can now precompute some
+            // common terms that are repeatedly applied in the inner loop.
+            // We won't be excessive here, but will rather focus on expensive
+            // operations, namely those involving the rank-4 material stiffness
+            // tensor and the rank-2 stress tensor.
+            //
+            // What we may observe is that both of these tensors are contracted
+            // with shape function gradients indexed on the "i" DoF. This
+            // implies that this particular operation remains constant as we
+            // loop over the "j" DoF. For that reason, we can extract this from
+            // the inner loop and save the many operations that, for each
+            // quadrature point and DoF index "i" and repeated over index "j"
+            // are required to double contract a rank-2 symmetric tensor with a
+            // rank-4 symmetric tensor, and a rank-1 tensor with a rank-2
+            // tensor.
+            //
+            // At the loss of some readability, this small change will reduce
+            // the assembly time of the symmetrized system by about half when
+            // using the simulation default parameters, and becomes more
+            // significant as the h-refinement level increases.
+            const SymmetricTensor<2, dim> symm_grad_Nx_i_x_Jc =
+              symm_grad_Nx[i] * Jc;
+            const Tensor<1, dim> grad_Nx_i_comp_i_x_tau_ns =
+              grad_Nx[i][component_i] * tau_ns;
```

Comment:
I guess this does not really pertain to this PR, but I wonder why you use separate arrays for `symm_grad_Nx` and `grad_Nx`, where the former is simply `symmetrize(grad_Nx)`? I would find the `symmetrize` approach clearer because I have to keep track of one variable less, but it might be just me.

Note that I am fine with the temporary variable you introduce, it's just the array coming from the outside.

Reply:
The simple answer is that I wanted to precompute as much as possible, and avoid the repeated call to `symmetrize(grad_Nx)` in the inner loop. I've got no concrete values to say how much quicker it is (if its quicker at all!) but it was just the philosophy I decided to apply right from the start and I stuck with it. Its true that its one more bit of bookkeeping to do, but that's one of several places in this tutorial where I'd gone for practicality or extensibility over the traditional or simple approach. 

Most of what's done here is, I guess, superseded by our fancy [`ScratchData`](https://www.dealii.org/9.2.0/doxygen/deal.II/classMeshWorker_1_1ScratchData.html) class, but in the past in other codes I would actually move the part that fills the cache somewhere else and share it between multiple weak forms.

> I guess this does not really pertain to this PR

I'd say it would have been a prime topic for discussion 10 years ago, when the tutorial was contributed :-D